### PR TITLE
PYTHON-3457 & PYTHON-3458 - Improve performance when logging is disabled

### PR DESCRIPTION
**Pull request by NoahStapp**

### CODE REVIEW

File: pymongo/network.py [hunk 1]
```diff
@@ -161,9 +163,26 @@ def command(
 
     if max_bson_size is not None and size > max_bson_size + message._COMMAND_OVERHEAD:
         message._raise_document_too_large(name, size, max_bson_size + message._COMMAND_OVERHEAD)
-
+    command_logger = logging.getLogger("pymongo.command")
+    # TODO: add serverConnectionId
```
Comment (ShaneHarvey) [inline pymongo/network.py:167]: Is this todo resolved? Or can you link the PYTHON ticket?

File: pymongo/message.py [hunk 2]
```diff
@@ -1001,33 +1003,84 @@ def unack_write(
         msg: bytes,
         max_doc_size: int,
         docs: list[Mapping[str, Any]],
+        client: MongoClient,
     ) -> Optional[Mapping[str, Any]]:
         """A proxy for Connection.unack_write that handles event publishing."""
+        command_logger = logging.getLogger("pymongo.command")
```
Comment (ShaneHarvey) [inline pymongo/message.py:1009]: Could you open a ticket to document how apps can silence and configure logging output? 

For example can they enable logging in thier own app but silence all pymongo logs? Do they need to silence each logger separately ("pymongo.command", "pymongo.pool", etc...)?

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

**Comment by ShaneHarvey** [comment]
I'm concerned with this approach because it results in 10% drop in single threaded performance *even when logging is not enabled*:
Screenshot 2024-01-25 at 10 54 05 AM
https://spruce.mongodb.com/task/mongo_python_driver_perf_tests_perf_6.0_standalone_patch_c4e4bd638f7b82032cbd13b8e56a7593b56e964c_65ab0a2c9ccd4ef5608366ab_24_01_19_23_47_57/trend-charts?execution=0&sortBy=STATUS&sortDir=ASC

How do we address this problem? To start we should check where this extra time is being spent. Is there a way to avoid this overhead when logging is not enabled? I'm concerned this drop in performance will get significantly worse when we add even more logging calls for topology, serverSelection, and connection.

File: pymongo/network.py (hunk 1 above)
Comment (NoahStapp) [reply pymongo/network.py:167]: Yes, resolved: https://jira.mongodb.org/browse/PYTHON-2822

File: test/unified_format.py [hunk 3]
```diff
@@ -629,7 +629,10 @@ def _operation_exists(self, spec, actual, key_to_compare):
             if key_to_compare is None:
                 assert actual is None
             else:
-                self.test.assertNotIn(key_to_compare, actual)
+                if key_to_compare in actual:
+                    assert actual[key_to_compare] is None
```
[the remark this answers is not in the document]
Comment (NoahStapp) [reply test/unified_format.py]: Good catch, removed.

File: test/test_custom_types.py [hunk 4]
```diff
@@ -92,6 +92,9 @@ def __eq__(self, other):
         # Does not compare equal to integers.
         return False
 
+    def __repr__(self):
```
[the remark this answers is not in the document]
Comment (NoahStapp) [reply test/test_custom_types.py]: Not sure, removed.

File: pymongo/message.py (hunk 2 above)
Comment (NoahStapp) [reply pymongo/message.py]: Done: [PYTHON-4167](https://jira.mongodb.org/browse/PYTHON-4167)

**Comment by ShaneHarvey** [comment]
Could you share a screenshot of the latest perf results?

**Comment by NoahStapp** [comment]
> Could you share a screenshot of the latest perf results?
Screenshot 2024-01-26 at 10 40 34 AM

Much better performance across the board

**Review by ShaneHarvey — APPROVED** [review]
Great LGTM!  Mind updating the PR title to something like "improve perf when logging is disabled"?

**Pull request merged**